module/zfs/arc.c

```diff
@@ -7983,7 +7983,12 @@ arc_init(void)
 	    offsetof(arc_prune_t, p_node));
 	mutex_init(&arc_prune_mtx, NULL, MUTEX_DEFAULT, NULL);
 
-	arc_prune_taskq = taskq_create("arc_prune", 100, defclsyspri,
+#ifdef __FreeBSD__
+	percent = 0;	/* More than one thread makes no sense. */
+#else
+	percent = 80;	/* Can use 1-2 threads per mount point. */
+#endif
```

Comment:
My understanding is that a single thread should be sufficient for Linux as well.  We can make that the default for both platforms.   Related to this change there's some related discussion in issue #9966.  I know @chrisrd has made some adjustments in this area and could comment more on this.

Reply:
Certainly a single thread improves my (Linux) workload by helping avoid `arc_prune` chewing up all the CPU. Others have seen the same (e.g. https://github.com/openzfs/zfs/issues/9966#issuecomment-955107413). I don't know if there are workloads which benefit from multiple threads, we could potentially make the number of threads configurable for Linux (e.g. see the `arc.c` changes in https://github.com/openzfs/zfs/files/7402413/zfs-prune-patch.txt), but I suspect setting it permanently to single threaded is fine.

Reply:
Another point is that "way back" it used to be unthreaded, i.e. running inline. Potentially it could be switched back, but that would require some detailed analysis of the locking to ensure there are no deadlocks due to lock inversions.

Reply:
> "way back" it used to be unthreaded

See commit f60467383655

Reply:
Issue #9966 was interesting reading.  It looks alike to what I was actually investigating on FreeBSD.  This patch is obviously not a fix for it in any way (except for FreeBSD multiple threads are waste even when not burning CPUs), but instead I am positioning it as a quick band-aid, allowing system to function while we diagnose the real vnodes reclamation problem.  If public opinion is that Linux does not need multiple threads also, I'll remake the patch that way tomorrow, making it as tunable in case somebody want more.  After this band-aid applied (I'd like it in 2.1 and even 2.0) I may find some time to look deeper into that part of arc reclamation.

Reply:
I'm happy with the quick band-aid going in.

I'm also interested in hearing more about the issue you're seeing on FreeBSD, e.g. is it the "arc_prune going nuts and chewing up 100%CPU" issue that's been around for a while?

Reply:
I have a problem report from customer, who periodically has system becoming unresponsive with all CPU time consumed by arc_prune threads.  But same as in #9966 I have no idea what triggers the problem, partially because it is difficult to collect information when it happen.

Reply:
If you're able, it would be great to see if the patch in https://github.com/openzfs/zfs/issues/9966#issuecomment-950110883 helps your client (minus the  `prune_no_aliases` stuff if you like - that's not necessary).

Reply:
Changed to single thread.

Reply:
@chrisrd I've looked on your patch, and it looks good as somewhat better Linux-specific workaround, but:
1) it won't help with actual ARC reclamation, only reduce CPU usage, and
2) the reclamation functions on FreeBSD do not return results, so it would require kernel patching first.